<a href="https://colab.research.google.com/github/Nniikkoollaass/data-science-modul-13-HW/blob/main/DS_HW_13.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Частина 1

​

В якості домашнього завдання вам пропонується створити нейронну мережу за допомогою механізмів Keras, яка буде класифікувати товари із датасету fasion_mnist.



На відміну від попереднього завдання вам пропонується створити згорткову нейромережу. Підберіть архітектуру мережі та навчіть її на даних із датасету fasion_mnist. Спробуйте досягти максимально можливої точності класифікації за рахунок маніпуляції параметрами мережі. Порівняйте точність отриманої згорткової мережі з точністю багатошарової мережі з попереднього завдання. Зробіть висновки.

In [ ]:
import tensorflow as tf
from tensorflow.keras.datasets import fashion_mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.optimizers import Adam

# Завантаження даних
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

# Зміна форми даних
train_images = train_images.reshape((60000, 28, 28, 1))
test_images = test_images.reshape((10000, 28, 28, 1))

# Нормалізація даних
train_images, test_images = train_images / 255.0, test_images / 255.0

# Перетворення міток у категоріальний формат
train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

# Створення моделі
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    MaxPooling2D((2, 2)),
    Conv2D(64, (3, 3), activation='relu'),
    MaxPooling2D((2, 2)),
    Conv2D(128, (3, 3), activation='relu'),
    Flatten(),
    Dense(128, activation='relu'),
    Dropout(0.5),
    Dense(10, activation='softmax')
])

# Компіляція моделі
model.compile(optimizer=Adam(),
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Навчання моделі
history = model.fit(train_images, train_labels, epochs=10, batch_size=64, validation_split=0.2)

# Оцінка моделі
test_loss, test_acc = model.evaluate(test_images, test_labels)
print(f'Test accuracy: {test_acc}')


4422102/4422102 [==============================] - 0s 0us/step
Epoch 1/10
750/750 [==============================] - 56s 71ms/step - loss: 0.6538 - accuracy: 0.7643 - val_loss: 0.3891 - val_accuracy: 0.8578
Epoch 2/10
750/750 [==============================] - 49s 66ms/step - loss: 0.4027 - accuracy: 0.8572 - val_loss: 0.3270 - val_accuracy: 0.8813
Epoch 3/10
750/750 [==============================] - 48s 65ms/step - loss: 0.3397 - accuracy: 0.8791 - val_loss: 0.3155 - val_accuracy: 0.8817
Epoch 4/10
750/750 [==============================] - 49s 65ms/step - loss: 0.3079 - accuracy: 0.8894 - val_loss: 0.2968 - val_accuracy: 0.8856
Epoch 5/10
750/750 [==============================] - 52s 69ms/step - loss: 0.2841 - accuracy: 0.8989 - val_loss: 0.2801 - val_accuracy: 0.8958
Epoch 6/10
750/750 [==============================] - 50s 67ms/step - loss: 0.2581 - accuracy: 0.9075 - val_loss: 0.2644 - val_accuracy: 0.9022
Epoch 7/10
750/750 [==============================] - 51s 68ms/step - los

Порівняння результатів багатошарової нейронної мережі (MLP) та згорткової нейронної мережі (CNN)



Результати багатошарової нейронної мережі (MLP)

Точність на валідаційному наборі даних (validation accuracy):
Початкова точність після першої епохи: 84.48%
Максимальна точність після навчання: 89.00%

Втрата на валідаційному наборі даних (validation loss):
Початкова втрата після першої епохи: 0.4337
Найменша втрата після навчання: 0.3041

Точність на тестовому наборі даних (test accuracy): 88.65%


Результати згорткової нейронної мережі (CNN)

Точність на валідаційному наборі даних (validation accuracy):
Початкова точність після першої епохи: 85.78%
Максимальна точність після навчання: 91.36%

Втрата на валідаційному наборі даних (validation loss):
Початкова втрата після першої епохи: 0.3891
Найменша втрата після навчання: 0.2435

Точність на тестовому наборі даних (test accuracy): 90.63%


Висновки

Точність моделі:

 Згорткова нейронна мережа (CNN) показала вищу точність на тестових даних (90.63%) у порівнянні з багатошаровою нейронною мережею (MLP) (88.65%). Це підтверджує, що CNN краще справляється з класифікацією зображень, ніж MLP, завдяки здатності виявляти просторові характеристики зображень.

Втрата:

CNN також показала меншу втрату на валідаційному наборі даних (0.2435) у порівнянні з MLP (0.3041). Це означає, що модель краще навчалася на тренувальних даних і має кращу здатність до узагальнення.

Переваги CNN:

Використання згорткових шарів у CNN дозволяє ефективніше витягувати ознаки з зображень, що робить цю архітектуру більш придатною для задач, пов'язаних з обробкою зображень.


Загальний висновок

Згорткова нейронна мережа (CNN) виявилася більш ефективною для задачі класифікації зображень з датасету Fashion MNIST, досягнувши вищої точності і меншої втрати в порівнянні з багатошаровою нейронною мережею (MLP). Це демонструє перевагу використання CNN для подібних задач і підтверджує її здатність краще витягувати і класифікувати просторові ознаки зображень.

Частина 2

​

В цій частині ми знову будемо працювати з датасетом fasion_mnist.



На відміну від попереднього завдання вам пропонується створити згорткову нейромережу, що використовує VGG16 в якості згорткової основи.



Навчіть отриману мережу на даних із датасету fasion_mnist. Спробуйте досягти максимально можливої точності класифікації за рахунок маніпуляції параметрами мережі. Під час навчання використовуйте прийоми донавчання та виділення ознак.



Порівняйте точність отриманої згорткової мережі з точністю багатошарової мережі з попереднього завдання. Зробіть висновки.

In [1]:
import tensorflow as tf
from tensorflow.keras.datasets import fashion_mnist
from tensorflow.keras.applications import MobileNetV2
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Dense, Flatten, Dropout, GlobalAveragePooling2D
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.utils import to_categorical
import numpy as np

# Завантаження датасету
(x_train, y_train), (x_test, y_test) = fashion_mnist.load_data()

# Змінення розмірів зображень до формату, необхідного для MobileNetV2
x_train = np.repeat(x_train[..., np.newaxis], 3, -1)  # Перетворення на RGB
x_test = np.repeat(x_test[..., np.newaxis], 3, -1)    # Перетворення на RGB

# Зміна розмірів зображень до 64x64, щоб ще більше зменшити використання пам'яті
x_train = tf.image.resize(x_train, (64, 64))
x_test = tf.image.resize(x_test, (64, 64))

# Нормалізація зображень
x_train = x_train / 255.0
x_test = x_test / 255.0

# Кодування міток
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# Завантаження попередньо натренованої моделі MobileNetV2 без верхніх шарів
mobilenet_v2_base = MobileNetV2(weights='imagenet', include_top=False, input_shape=(64, 64, 3))

# Замороження шарів базової моделі
for layer in mobilenet_v2_base.layers:
    layer.trainable = False

# Додавання нових шарів до базової моделі
x = GlobalAveragePooling2D()(mobilenet_v2_base.output)
x = Dense(128, activation='relu')(x)
x = Dropout(0.5)(x)
output = Dense(10, activation='softmax')(x)

# Створення нової моделі
model = Model(inputs=mobilenet_v2_base.input, outputs=output)

# Компіляція моделі
model.compile(optimizer=Adam(), loss='categorical_crossentropy', metrics=['accuracy'])

# Навчання моделі з ще меншим batch size
history = model.fit(x_train, y_train, epochs=10, batch_size=16, validation_split=0.2)

# Оцінка моделі на тестовому наборі даних
test_loss, test_accuracy = model.evaluate(x_test, y_test)
print(f'Test accuracy: {test_accuracy}')


9406464/9406464 [==============================] - 0s 0us/step
Epoch 1/10
3000/3000 [==============================] - 198s 65ms/step - loss: 0.6159 - accuracy: 0.7869 - val_loss: 0.4253 - val_accuracy: 0.8461
Epoch 2/10
3000/3000 [==============================] - 186s 62ms/step - loss: 0.4687 - accuracy: 0.8315 - val_loss: 0.4111 - val_accuracy: 0.8472
Epoch 3/10
3000/3000 [==============================] - 200s 67ms/step - loss: 0.4327 - accuracy: 0.8441 - val_loss: 0.3967 - val_accuracy: 0.8562
Epoch 4/10
3000/3000 [==============================] - 194s 65ms/step - loss: 0.4092 - accuracy: 0.8507 - val_loss: 0.3916 - val_accuracy: 0.8602
Epoch 5/10
3000/3000 [==============================] - 194s 65ms/step - loss: 0.3849 - accuracy: 0.8591 - val_loss: 0.3854 - val_accuracy: 0.8648
Epoch 6/10
3000/3000 [==============================] - 192s 64ms/step - loss: 0.3681 - accuracy: 0.8646 - val_loss: 0.3941 - val_accuracy: 0.8582
Epoch 7/10
3000/3000 [==============================] -

Висновки

Точність:
Тестова точність багатошарової нейронної мережі становить 88.65%, тоді як тестова точність мережі на основі MobileNetV2 становить 86.29%. Багатошарова нейронна мережа показала кращу точність на тестовому наборі даних.

Швидкість навчання:
MobileNetV2 навчалася протягом 10 епох з кожною епохою, що займає приблизно 194-200 секунд.
Багатошарова нейронна мережа навчалася протягом 24 епох з кожною епохою, що займає приблизно 2-5 секунд.
MobileNetV2 потребує більше часу на навчання за епоху, але завершила навчання всього за 10 епох.

Використання пам'яті:
MobileNetV2 зменшила використання пам'яті завдяки зменшенню розміру зображень і менш складній моделі. Однак вона все одно споживає більше пам'яті, ніж багатошарова нейронна мережа, через свою складнішу архітектуру.

Втрати:
Багатошарова нейронна мережа показує стабільніше зниження втрат (loss) протягом епох порівняно з MobileNetV2, де втрати залишаються на вищому рівні, особливо під час валідації.

Узагальнення:
Якщо основний пріоритет — точність, багатошарова нейронна мережа є кращим варіантом для цього завдання.
Якщо необхідно зменшити використання пам'яті і прийнятна менша точність, MobileNetV2 може бути підходящим вибором.